File: src/main/kotlin/com/linqingying/cangjie/psi/CjMatchEntry.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.lexer.CjTokens;
import com.linqingying.cangjie.psi.psiUtil.CjPsiUtilKt;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


public class CjMatchEntry extends CjElementImpl {
    public CjMatchEntry(@NotNull ASTNode node) {
        super(node);
    }

    public boolean is_() {
        return get_Keyword() != null;
    }

    public PsiElement get_Keyword() {
        return findChildByType(CjTokens.UNDERLINE);
    }

    @Nullable
    public CjExpression getExpression() {
        return findChildByClass(CjExpression.class);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitMatchEntry(this, data);
    }

    @NotNull
    public CjMatchCondition[] getConditions() {
        return findChildrenByClass(CjMatchCondition.class);
    }

    public PsiElement getTrailingComma() {
        return CjPsiUtilKt.getTrailingCommaByClosingElement(getArrow());
    }

    @Nullable
    public PsiElement getArrow() {
        return findChildByType(CjTokens.DOUBLE_ARROW);
    }
}

File: src/main/kotlin/com/huawei/cangjie/psi/CjMatchEntry.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.CjNodeTypes
import com.huawei.cangjie.lexer.CjTokens
import com.huawei.cangjie.psi.psiUtil.getTrailingCommaByClosingElement
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement

interface CjPatternEntryBlock:PsiElement

class CjMatchEntry(node: ASTNode) : CjElementImpl(node),CjPatternEntryBlock {
    val isElse: Boolean
        get() {
            return elseKeyword != null
        }

    val elseKeyword: PsiElement?
        get() {
            return findChildByType(CjNodeTypes.WILDCARD_PATTERN)
        }

    val expression: CjCaseBlockExpression?
        get() = findChildByClass(CjCaseBlockExpression::class.java)

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitMatchEntry(this, data)
    }

    val conditions: Array<CjCasePattern>
        get() = findChildrenByClass(CjCasePattern::class.java)

    val trailingComma: PsiElement?
        get() = getTrailingCommaByClosingElement(arrow)

    val arrow: PsiElement?
        get() = findChildByType(CjTokens.DOUBLE_ARROW)
}
